File: src/typescript/mount.ts
/**
 * Programmatic mount sources for the virtual filesystem.
 *
 * A {@link MountSource} exposes a live, read-only directory tree of unknown
 * shape (a host folder, a GitHub repo, Slack messages, ...). The
 * {@link MountingFilesystemDriver} overlays one or more sources onto an inner
 * writable {@link FilesystemDriver} using copy-up semantics: writes shadow the
 * source in the inner layer (the source is never mutated), and deleting a
 * source-only path throws (read-only, no whiteout/tombstone in v1).
 *
 * See docs/adr/0034-programmatic-mount-sources.md.
 */

import { VirtualFS } from "./virtual-fs.js";
import type { FilesystemDriver } from "./drivers.js";

/** Lightweight stat for a path inside a mount source. */
export interface MountStat {
  isDir: boolean;
  size?: number;
  mtime?: number;
}

/**
 * Read-only provider of a directory tree, addressed by relative subpaths
 * ("" is the mount root). `stat` is called constantly for existence checks and
 * must return `null` for missing paths (no throwing for control flow);
 * `list`/`read` are only called after `stat` confirms existence.
 */
export interface MountSource {
  stat(subpath: string): MountStat | null;
  list(subpath: string): string[];
  read(subpath: string): string;
}

/** Capability interface for filesystems that support mounting sources. */
export interface Mountable {
  mount(mountPoint: string, source: MountSource): void;
  unmount(mountPoint: string): void;
  mounts(): string[];
}

/** Type guard for {@link Mountable}. */
export function isMountable(fs: unknown): fs is Mountable {
  return (
    typeof fs === "object" &&
    fs !== null &&
    typeof (fs as Mountable).mount === "function" &&
    typeof (fs as Mountable).unmount === "function" &&
    typeof (fs as Mountable).mounts === "function"
  );
}

const norm = (p: string): string => VirtualFS._norm(p);

interface MountEntry {
  point: string;
  source: MountSource;
}

/**
 * Overlays read-only {@link MountSource}s onto an inner writable driver.
 * Empty mount table => every method is a transparent passthrough to inner.
 */
export class MountingFilesystemDriver implements FilesystemDriver, Mountable {
  private _inner: FilesystemDriver;
  private _mounts: MountEntry[];

  constructor(inner: FilesystemDriver, mounts?: MountEntry[]) {
    this._inner = inner;
    this._mounts = mounts ? [...mounts] : [];
  }

  // --- Mountable ----------------------------------------------------------

  mount(mountPoint: string, source: MountSource): void {
    mountPoint = norm(mountPoint);
    this.unmount(mountPoint);
    this._mounts.push({ point: mountPoint, source });
  }

  unmount(mountPoint: string): void {
    mountPoint = norm(mountPoint);
    this._mounts = this._mounts.filter((e) => e.point !== mountPoint);
  }

  mounts(): string[] {
    return this._mounts.map((e) => e.point).sort();
  }

  // --- routing helpers ----------------------------------------------------

  private _resolveMount(p: string): { source: MountSource; subpath: string } | null {
    p = norm(p);
    let best: { point: string; source: MountSource; subpath: string } | null = null;
    for (const { point, source } of this._mounts) {
      let subpath: string;
      if (p === point) subpath = "";
      else if (point === "/") subpath = p.slice(1);
      else if (p.startsWith(point + "/")) subpath = p.slice(point.length + 1);
      else continue;
      if (best === null || point.length > best.point.length) {
        best = { point, source, subpath };
      }
    }
    return best ? { source: best.source, subpath: best.subpath } : null;
  }

  private _isMountPointOrAncestor(p: string): boolean {
    p = norm(p);
    const prefix = p === "/" ? "/" : p + "/";
    return this._mounts.some((e) => e.point === p || e.point.startsWith(prefix));
  }

  // --- FilesystemDriver ---------------------------------------------------

  write(path: string, content: string): void {
    this._inner.write(path, content);
  }

  writeLazy(path: string, provider: () => string): void {
    this._inner.writeLazy(path, provider);
  }

  read(path: string): string {
    if (this._inner.exists(path) && !this._inner.isDir(path)) {
      return this._inner.read(path);
    }
    const m = this._resolveMount(path);
    if (m) {
      const st = m.source.stat(m.subpath);
      if (st !== null && !st.isDir) return m.source.read(m.subpath);
    }
    return this._inner.read(path); // throws "No such file"
  }

  readText(path: string): string {
    return this.read(path);
  }

  exists(path: string): boolean {
    if (this._inner.exists(path)) return true;
    if (this._isMountPointOrAncestor(path)) return true;
    const m = this._resolveMount(path);
    return m !== null && m.source.stat(m.subpath) !== null;
  }

  remove(path: string): void {
    if (this._inner.exists(path)) {
      this._inner.remove(path);
      return;
    }
    const m = this._resolveMount(path);
    if (m && m.source.stat(m.subpath) !== null) {
      throw new Error(`${norm(path)}: read-only mount source (cannot delete)`);
    }
    throw new Error(`${norm(path)}: No such file`);
  }

  isDir(path: string): boolean {
    if (this._inner.isDir(path)) return true;
    if (this._isMountPointOrAncestor(path)) return true;
    const m = this._resolveMount(path);
    if (m) {
      const st = m.source.stat(m.subpath);
      return st !== null && st.isDir;
    }
    return false;
  }

  // private alias used by the shell builtins (parity with VirtualFS)
  _isDir(path: string): boolean {
    return this.isDir(path);
  }

  listdir(path: string = "/"): string[] {
    path = norm(path);
    const entries = new Set<string>(this._inner.listdir(path));
    const m = this._resolveMount(path);
    if (m) {
      const st = m.source.stat(m.subpath);
      if (st !== null && st.isDir) {
        for (const name of m.source.list(m.subpath)) entries.add(name);
      }
    }
    const prefix = path === "/" ? "/" : path + "/";
    for (const e of this._mounts) {
      if (e.point.startsWith(prefix)) {
        const seg = e.point.slice(prefix.length).split("/")[0];
        if (seg) entries.add(seg);
      }
    }
    return [...entries].sort();
  }

  find(root: string = "/", pattern: string = "*"): string[] {
    root = norm(root);
    const regex = MountingFilesystemDriver._globToRegex(pattern);
    const results: string[] = [];
    const walk = (dir: string): void => {
      for (const name of this.listdir(dir)) {
        const full = dir === "/" ? "/" + name : dir + "/" + name;
        const isDir = this.isDir(full);
        if (!isDir && regex.test(name)) results.push(full);
        if (isDir) walk(full);
      }
    };
    if (this.isDir(root)) walk(root);
    return results.sort();
  }

  private static _globToRegex(pattern: string): RegExp {
    let regexStr = "^";
    for (const ch of pattern) {
      if (ch === "*") regexStr += ".*";
      else if (ch === "?") regexStr += ".";
      else if (".+^${}()|[]\\".includes(ch)) regexStr += "\\" + ch;
      else regexStr += ch;
    }
    return new RegExp(regexStr + "$");
  }

  stat(path: string): { path: string; type: string; size?: number; mtime?: number } {
    if (this._inner.exists(path)) return this._inner.stat(path);
    const np = norm(path);
    if (this._isMountPointOrAncestor(np)) return { path: np, type: "directory" };
    const m = this._resolveMount(np);
    if (m) {
      const st = m.source.stat(m.subpath);
      if (st !== null) {
        if (st.isDir) return { path: np, type: "directory" };
        return { path: np, type: "file", size: st.size ?? 0, mtime: st.mtime ?? 0 };
      }
    }
    return this._inner.stat(path); // throws "No such file"
  }

  clone(): MountingFilesystemDriver {
    // Sources are read-only and shareable; copy the table by reference.
    return new MountingFilesystemDriver(this._inner.clone(), this._mounts);
  }
}

File: src/python/mount.py
"""Programmatic mount sources for the virtual filesystem.

A ``MountSource`` exposes a live, read-only directory tree of unknown shape
(a host folder, a GitHub repo, Slack messages, a DB table, ...). The
``MountingFilesystemDriver`` overlays one or more sources onto an inner
writable ``FilesystemDriver`` using copy-up semantics: writes shadow the
source in the inner layer (the source is never mutated), and deleting a
source-only path raises (read-only, no whiteout/tombstone in v1).

See docs/adr/0034-programmatic-mount-sources.md.
"""

from __future__ import annotations

import fnmatch
import os
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any, Callable, Protocol, runtime_checkable

from src.python.drivers import FilesystemDriver


@dataclass
class MountStat:
    """Lightweight stat for a path inside a mount source."""

    is_dir: bool
    size: int = 0
    mtime: int = 0


class MountSource(ABC):
    """Read-only provider of a directory tree, addressed by relative subpaths.

    ``""`` is the mount root. ``stat`` is called constantly to check existence
    and must not raise for missing paths (return ``None``). ``list``/``read``
    are only called after ``stat`` confirms existence and may raise on real
    I/O faults.
    """

    @abstractmethod
    def stat(self, subpath: str) -> MountStat | None: ...

    @abstractmethod
    def list(self, subpath: str) -> list[str]: ...

    @abstractmethod
    def read(self, subpath: str) -> str | bytes: ...


@runtime_checkable
class Mountable(Protocol):
    """Capability interface for filesystems that support mounting sources."""

    def mount(self, mount_point: str, source: MountSource) -> None: ...

    def unmount(self, mount_point: str) -> None: ...

    def mounts(self) -> list[str]: ...


def _norm(path: str) -> str:
    result = os.path.normpath("/" + path)
    if result.startswith("//"):
        result = result[1:]
    return result


class MountingFilesystemDriver(FilesystemDriver):
    """Overlays read-only ``MountSource``s on an inner writable driver.

    Empty mount table => every method is a transparent passthrough to inner.
    """

    def __init__(
        self,
        inner: FilesystemDriver,
        mounts: list[tuple[str, MountSource]] | None = None,
    ):
        self._inner = inner
        # Ordered mount table: (mount_point, source).
        self._mounts: list[tuple[str, MountSource]] = list(mounts) if mounts else []

    # -- Mountable capability ------------------------------------------------

    def mount(self, mount_point: str, source: MountSource) -> None:
        mount_point = _norm(mount_point)
        self.unmount(mount_point)
        self._mounts.append((mount_point, source))

    def unmount(self, mount_point: str) -> None:
        mount_point = _norm(mount_point)
        self._mounts = [(m, s) for m, s in self._mounts if m != mount_point]

    def mounts(self) -> list[str]:
        return sorted(m for m, _ in self._mounts)

    # -- Routing helpers -----------------------------------------------------

    def _resolve_mount(self, path: str) -> tuple[str, MountSource, str] | None:
        """Longest matching mount prefix -> (mount_point, source, subpath)."""
        path = _norm(path)
        best: tuple[str, MountSource, str] | None = None
        for mount_point, source in self._mounts:
            if path == mount_point:
                subpath = ""
            elif mount_point == "/":
                subpath = path[1:]
            elif path.startswith(mount_point + "/"):
                subpath = path[len(mount_point) + 1:]
            else:
                continue
            if best is None or len(mount_point) > len(best[0]):
                best = (mount_point, source, subpath)
        return best

    def _is_mount_point_or_ancestor(self, path: str) -> bool:
        path = _norm(path)
        prefix = "/" if path == "/" else path + "/"
        for mount_point, _ in self._mounts:
            if mount_point == path or mount_point.startswith(prefix):
                return True
        return False

    # -- FilesystemDriver ----------------------------------------------------

    def write(self, path: str, content: str | bytes) -> None:
        self._inner.write(path, content)

    def write_lazy(self, path: str, provider: Callable[[], str | bytes]) -> None:
        self._inner.write_lazy(path, provider)

    def read(self, path: str) -> str | bytes:
        if self._inner.exists(path) and not self._inner.is_dir(path):
            return self._inner.read(path)
        resolved = self._resolve_mount(path)
        if resolved is not None:
            _, source, subpath = resolved
            st = source.stat(subpath)
            if st is not None and not st.is_dir:
                return source.read(subpath)
        return self._inner.read(path)  # raises FileNotFoundError

    def read_text(self, path: str) -> str:
        content = self.read(path)
        return content if isinstance(content, str) else content.decode(
            "utf-8", errors="replace"
        )

    def exists(self, path: str) -> bool:
        if self._inner.exists(path):
            return True
        if self._is_mount_point_or_ancestor(path):
            return True
        resolved = self._resolve_mount(path)
        if resolved is not None:
            _, source, subpath = resolved
            return source.stat(subpath) is not None
        return False

    def remove(self, path: str) -> None:
        if self._inner.exists(path):
            self._inner.remove(path)
            return
        resolved = self._resolve_mount(path)
        if resolved is not None:
            _, source, subpath = resolved
            if source.stat(subpath) is not None:
                raise PermissionError(
                    f"{_norm(path)}: read-only mount source (cannot delete)"
                )
        raise FileNotFoundError(f"{_norm(path)}: No such file")

    def is_dir(self, path: str) -> bool:
        if self._inner.is_dir(path):
            return True
        if self._is_mount_point_or_ancestor(path):
            return True
        resolved = self._resolve_mount(path)
        if resolved is not None:
            _, source, subpath = resolved
            st = source.stat(subpath)
            return st is not None and st.is_dir
        return False

    # private alias used by the shell builtins (parity with VirtualFS)
    _is_dir = is_dir

    def listdir(self, path: str = "/") -> list[str]:
        path = _norm(path)
        entries: set[str] = set(self._inner.listdir(path))
        # Source children when this path is a directory inside a mount.
        resolved = self._resolve_mount(path)
        if resolved is not None:
            _, source, subpath = resolved
            st = source.stat(subpath)
            if st is not None and st.is_dir:
                entries.update(source.list(subpath))
        # Mount points whose first segment is rooted at this directory.
        prefix = "/" if path == "/" else path + "/"
        for mount_point, _ in self._mounts:
            if mount_point.startswith(prefix):
                entries.add(mount_point[len(prefix):].split("/")[0])
        return sorted(entries)

    def find(self, root: str = "/", pattern: str = "*") -> list[str]:
        root = _norm(root)
        results: list[str] = []

        def walk(directory: str) -> None:
            for name in self.listdir(directory):
                full = (directory.rstrip("/") + "/" + name) if directory != "/" else "/" + name
                is_dir = self.is_dir(full)
                if not is_dir and fnmatch.fnmatch(name, pattern):
                    results.append(full)
                if is_dir:
                    walk(full)

        if self.is_dir(root):
            walk(root)
        return sorted(results)

    def stat(self, path: str) -> dict[str, Any]:
        if self._inner.exists(path):
            return self._inner.stat(path)
        npath = _norm(path)
        if self._is_mount_point_or_ancestor(npath):
            return {"path": npath, "type": "directory"}
        resolved = self._resolve_mount(npath)
        if resolved is not None:
            _, source, subpath = resolved
            st = source.stat(subpath)
            if st is not None:
                if st.is_dir:
                    return {"path": npath, "type": "directory"}
                return {
                    "path": npath,
                    "type": "file",
                    "size": st.size,
                    "mtime": st.mtime,
                }
        return self._inner.stat(path)  # raises FileNotFoundError

    def clone(self) -> MountingFilesystemDriver:
        # Sources are read-only and shareable; copy the table by reference.
        return MountingFilesystemDriver(self._inner.clone(), list(self._mounts))
